fdbbackup/tests/s3_backup_test.sh

```diff
@@ -0,0 +1,251 @@
+    rm -rf "${SCRATCH_DIR}"
+  fi
+}
+
+# Resolve passed in reference to an absolute path.
+# e.g. /tmp on mac is actually /private/tmp.
+# $1 path to resolve
+function resolve_to_absolute_path {
+  local p="${1}"
+  while [[ -h "${p}" ]]; do
+    dir=$( cd -P "$( dirname "${p}" )" >/dev/null 2>&1 && pwd )
+    p=$(readlink "${p}")
+    [[ ${p} != /* ]] && p="${dir}/${p}"
+  done
+  realpath "${p}"
+}
+
+# Run the fdbbackup command.
+# $1 The build directory
+# $2 The scratch directory
+# $3 The weed port s3 is listening on.
+function backup {
+  local local_build_dir="${1}"
+  local scratch_dir="${2}"
+  local weed_s3_port="${3}"
+  # Backup to s3. Without the -k argument in the below, the backup gets
+  # 'No restore target version given, will use maximum restorable version from backup description.'
+  # TODO: Why is -k needed?
+  if ! "${local_build_dir}"/bin/fdbbackup start \
+    -C "${scratch_dir}/loopback_cluster/fdb.cluster" \
+    -t "${TAG}" -w \
+    -d "blobstore://localhost:${weed_s3_port}/${S3_RESOURCE}?bucket=${BUCKET}&secure_connection=0&region=us" \
+    -k '"" \xff' \
+    --log --logdir="${scratch_dir}" \
+    --knob_http_verbose_level=10 \
+    --knob_http_request_aws_v4_header=true
+  then
+    err "Start fdbbackup failed"
+    exit 1
+  fi
+}
+
+# Run the fdbrestore command.
+# $1 The build directory
+# $2 The scratch directory
+# $3 The weed port s3 is listening on.
+function restore {
+  local local_build_dir="${1}"
+  local scratch_dir="${2}"
+  local weed_s3_port="${3}"
+  if ! "${local_build_dir}"/bin/fdbrestore start \
+    --dest-cluster-file "${scratch_dir}/loopback_cluster/fdb.cluster" \
+    -t "${TAG}" -w \
+    -r "blobstore://localhost:${weed_s3_port}/${S3_RESOURCE}?bucket=${BUCKET}&secure_connection=0&region=us" \
+    --log --logdir="${scratch_dir}" \
+    --knob_http_verbose_level=10 -t first_backup  \
+    --knob_http_request_aws_v4_header=true
+  then
+    err "Start fdbrestore failed"
+    exit 1
+  fi
+}
+
+# Run a backup to s3 and then a restore.
+# $1 build directory
+# $2 the scratch directory
+# $3 the s3_port to go against.
+function test_s3_backup_and_restore {
+  local local_build_dir="${1}"
+  local scratch_dir="${2}"
+  local local_s3_port="${3}"
+  log "Load data"
+  if ! load_data "${local_build_dir}" "${scratch_dir}"; then
+    err "Failed loading data into fdb"
+    exit 1
+  fi
+  log "Run s3 backup"
+  if ! backup "${local_build_dir}" "${scratch_dir}" "${local_s3_port}"; then
+    err "Failed backup"
+    exit 1
+  fi
+  log "Clear fdb data"
+  if ! clear_data "${local_build_dir}" "${scratch_dir}"; then
+    err "Failed clear data in fdb"
+    exit 1
+  fi
+  log "Restore from s3"
+  if ! restore "${local_build_dir}" "${scratch_dir}" "${local_s3_port}"; then
+    err "Failed restore"
+    exit 1
+  fi
+  log "Verify restore"
+  if ! verify_data "${local_build_dir}" "${scratch_dir}"; then
+    err "Failed verification of data in fdb"
+    exit 1
+  fi
+}
```

Comment:
Can we add one more check here? I.e., `grep` the logs to see if there are `Severity="40"` events and fail the test if there are.

Reply:
Good idea. Let me push a change.

Reply:
I changed the code to look for 'Severity=30' and reran the tests. Looks like this when it fails...

```
98.2024-11-21T19:21:18-08:00
99.2024-11-21T19:21:18-08:00
2024-11-21T19:21:59-08:00 Check for Severity=40 errors
./s3_backup_test.sh: line 147: grep_for_severity40: command not found
2024-11-21T19:21:59-08:00 ERROR: Found Severity=40 errors in logs
./s3_backup_test.sh: line 46: 24592 Killed: 9               "${local_build_dir}/bin/backup_agent" -C "${scratch_dir}/loopback_cluster/fdb.cluster" --log --logdir="${scratch_dir}" --knob_http_verbose_level=10
```

Let me make it better... have it print out the lines from the log... Will be back.